### PR TITLE
Changes for MCQ

### PR DESCRIPTION
**Pull request by rakeshqdsa**

### CODE REVIEW

**Comment by rakeshqdsa** [comment]
https://learnosity.atlassian.net/browse/WBL-5

**Comment by rakeshqdsa** [comment]
https://learnosity.atlassian.net/browse/WBL-3

File: src/Converter.php [hunk 1]
```diff
@@ -27,9 +26,9 @@
 
 class Converter
 {
+
```
Comment (ttton) [inline src/Converter.php]: There are a lot of unnecessary whitespace/newline changes in this file. This affects readability, increases the surface area of the PR and makes reverts/cherry-picks harder - can we please reduce these to what's necessary?

File: src/Services/ConvertToQtiService.php [hunk 2]
```diff
@@ -169,18 +159,20 @@ private function convertAssessmentItem($json)
...
-        } else {
-            $result = [
-                '',
-                [
-                    'Ignoring ' . $json['type'] . ' , currently unsupported'
-                ]
-            ];
-            $this->output->writeln("  <error>Question type `{$json['type']}` not yet supported, ignoring</error>");
-        }
+        foreach ($json['questions'] as $question):
```
Comment (ttton) [inline src/Services/ConvertToQtiService.php]: Style: This works, but can we please use consistent syntax throughout the codebase? Colon control structures are not used anywhere else, so it is highly preferable we stick with brace syntax here. It also doesn't make a lot of sense that we're mixing it with brace syntax in the same block.

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php [hunk 3]
```diff
@@ -12,7 +14,7 @@ abstract class AbstractQuestionValidationBuilder
 
     abstract protected function buildResponseDeclaration($responseIdentifier, $validation);
 
-    public function buildValidation($responseIdentifier, $validation, $isCaseSensitive = true)
+    public function buildValidation($responseIdentifier, $validation, $distractor_rationale_response_level = array(), $isCaseSensitive = true)
```
Comment (ttton) [inline src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: Style: Stick to using lower-camel case for local and class variable names (rather than snake case) - i.e. `$distractorRationaleResponseLevel` instead of `$distractor_rationale_response_level`

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php [hunk 4]
```diff
@@ -3,7 +3,9 @@
 namespace LearnosityQti\Processors\QtiV2\Out\Validation;
 
 use LearnosityQti\Processors\QtiV2\Out\Constants;
+use LearnosityQti\Processors\QtiV2\Out\ResponseProcessingBuilders\QtiResponseProcessingBuilder;
 use LearnosityQti\Services\LogService;
+use LearnosityQti\Utils\QtiMarshallerUtil;
```
Comment (ttton) [inline src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: Remove this unused use declaration.

File: src/Services/ConvertToQtiService.php [hunk 5]
```diff
@@ -38,15 +38,13 @@ class ConvertToQtiService
     protected $shouldAppendLogs           = false;
     protected $shouldGuessItemScoringType = true;
     protected $shouldUseManifest          = true;
-
```
Comment (ttton) [inline src/Services/ConvertToQtiService.php]: Formatting: A lot of unnecessary whitespace changes in this file. This is likely to cause unnecessary merge conflicts, and doesn't improve readability. Can these be removed if they are not needed?

File: src/Processors/QtiV2/Out/ResponseProcessingBuilders/QtiResponseProcessingBuilder.php [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    public function build($score)
+    {
+        // creating feedback outcome
+        $feedbackResponseComponent = new SetOutcomeValue('FEEDBACK', new Variable('RESPONSE'));
+        $responseRuleCollection = new ResponseRuleCollection();
+        $responseRuleCollection->attach($feedbackResponseComponent);
+
+        // creating responseIf condition
+        $responseIfexpressionCollection = new ExpressionCollection();
```
Comment (ttton) [inline src/Processors/QtiV2/Out/ResponseProcessingBuilders/QtiResponseProcessingBuilder.php]: Style: Apply capitalization of camel case variable names consistently: i.e. `$responseIfExpressionCollection` instead of `$responseIfexpressionCollection`

File: src/Converter.php [hunk 7]
```diff
@@ -1,7 +1,5 @@
 <?php
-
 namespace LearnosityQti;
-
```
[the remark this answers is not in the document]
Comment (pooja31chaudhary) [reply src/Converter.php]: Already covered this issue. Please check the latest commit.

File: src/Converter.php [hunk 8]
```diff
@@ -24,16 +22,13 @@
 use qtism\data\storage\xml\XmlStorageException;
 use Symfony\Component\Finder\Finder;
 use Symfony\Component\Finder\SplFileInfo;
-
```
[the remark this answers is not in the document]
Comment (pooja31chaudhary) [reply src/Converter.php]: Already fixed this issues. Please check latest commit.

File: src/Converter.php [hunk 9]
```diff
@@ -41,15 +40,11 @@ public static function convertImscpDirectoryToLearnosityDirectory($imscpDirector
         $learnosityImagesDirectory = $learnosityDirectory . '/Images';
         FileSystemUtil::createOrReplaceDir($learnosityImagesDirectory);
         $allImages = array_merge(
-            self::extractFiles('*.jpg', $imscpDirectory, $learnosityImagesDirectory),
-            self::extractFiles('*.jpeg', $imscpDirectory, $learnosityImagesDirectory),
-            self::extractFiles('*.gif', $imscpDirectory, $learnosityImagesDirectory)
+            self::extractFiles('*.jpg', $imscpDirectory, $learnosityImagesDirectory), self::extractFiles('*.jpeg', $imscpDirectory, $learnosityImagesDirectory), self::extractFiles('*.gif', $imscpDirectory, $learnosityImagesDirectory)
```
[the remark this answers is not in the document]
Comment (pooja31chaudhary) [reply src/Converter.php]: Already fixed this issues. Please check latest commit.

File: src/Converter.php [hunk 10]
```diff
@@ -127,7 +119,6 @@ public static function convertQtiTestToLearnosity($xmlString, $validate = true)
     {
         $testMapper = AppContainer::getApplicationContainer()->get('qtiv2_test_mapper');
         $activityWriter = AppContainer::getApplicationContainer()->get('learnosity_activity_writer');
-
```
[the remark this answers is not in the document]
Comment (pooja31chaudhary) [reply src/Converter.php]: Already fixed this issues. Please check latest commit.

File: src/Converter.php (hunk 1 above)
Comment (pooja31chaudhary) [reply src/Converter.php]: Done

File: src/Services/ConvertToQtiService.php (hunk 2 above)
Comment (pooja31chaudhary) [reply src/Services/ConvertToQtiService.php]: Done

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php (hunk 3 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: Done.

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php (hunk 4 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: Done

File: src/Services/ConvertToQtiService.php (hunk 5 above)
Comment (pooja31chaudhary) [reply src/Services/ConvertToQtiService.php]: Done

File: src/Processors/QtiV2/Out/ResponseProcessingBuilders/QtiResponseProcessingBuilder.php (hunk 6 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/ResponseProcessingBuilders/QtiResponseProcessingBuilder.php]: Done

File: src/Converter.php [hunk 11]
```diff
@@ -310,10 +309,9 @@ private static function convertLearnosityQuestionData(array $questionDataJson)
     private static function convertLearnosityItem(array $itemJson)
     {
         // Separate question(s) and item
-        $itemJson['questionReferences'] = array_column($itemJson['questions'], 'reference');
+        $itemJson['questionReferences'] = array_column($itemJson['questions'], 'response_id');
```
Comment (ttton) [inline src/Converter.php]: I'm just curious about this particular change - it doesn't really make sense to me for the response ID to be passed to `questionReferences` (which to me should relate to the `reference` key of a question or feature). Why was this decided on?

Comment (rakeshqdsa) [reply src/Converter.php]: Earlier we were working on wrong item json where the key response_id was present instead of reference . But later Michael told us that we need to work in the data coming from offlinpackage api.
Now we have changed it to reference .

File: src/Converter.php [hunk 12]
```diff
@@ -325,6 +323,8 @@ private static function convertLearnosityItem(array $itemJson)
         $item = $itemMapper->parse($itemJson);
         $questions = [];
         foreach ($questionsJson as $question) {
+            $question['reference'] = $question['response_id'];
```
Comment (ttton) [inline src/Converter.php]: What's this change needed for? Is this also related to the missing `reference` data that you mentioned here?
https://github.com/Learnosity/learnosity-qti/pull/34#discussion_r295651473

File: src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php [hunk 13]
```diff
@@ -9,15 +9,21 @@
 use LearnosityQti\Entities\QuestionTypes\mcq_options_item;
 use LearnosityQti\Services\LogService;
 use LearnosityQti\Utils\QtiMarshallerUtil;
+use qtism\common\utils\Version;
 use qtism\common\utils\Format;
 use qtism\data\content\FlowStaticCollection;
+use qtism\data\content\FeedbackInline;
+use qtism\data\content\InlineCollection;
+use qtism\data\content\SimpleInline;
```
Comment (ttton) [inline src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php]: Is this used anywhere?

File: src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php [hunk 14]
```diff
@@ -26,11 +32,30 @@ public function convert(BaseQuestionType $questionType, $interactionIdentifier,
...
+        }
+        $i = 0;
         foreach ($question->get_options() as $index => $option) {
             /** @var mcq_options_item $option */
             $choiceContent = new FlowStaticCollection();
             foreach (QtiMarshallerUtil::unmarshallElement($option->get_label()) as $component) {
+
                 $choiceContent->attach($component);
+                // attach feedbackInline to simpleChoice
+                if (isset($feedbackOptions) && $feedbackOptions[$i] != '') {
```
Comment (ttton) [inline src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php]: Use identity comparison `!==` to avoid type juggling issues.

File: src/Services/ConvertToQtiService.php [hunk 15]
```diff
@@ -229,12 +210,14 @@ private function persistResultsFile(array $results, $outputFilePath)
...
             return;
         }
-
         $this->output->writeln("\n<info>" . static::INFO_OUTPUT_PREFIX . "Writing conversion results: " . $outputFilePath . '.json' . "</info>\n");
 
         foreach ($results as $result) {
-            if (!empty($result['qti'])) {
-                file_put_contents($outputFilePath . '/' . $result['json']['reference'] . '.xml', $result['qti'][0]);
+
+            foreach ($result['json']['questions'] as $question) {
```
Comment (ttton) [inline src/Services/ConvertToQtiService.php]: Minor one - you could make the `if` check wrap the `foreach` block instead, since it will return the same result for every iteration and you could avoid entering the loop at all if it resolves to `false`.

File: src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php [hunk 16]
```diff
@@ -72,7 +97,11 @@ public function convert(BaseQuestionType $questionType, $interactionIdentifier,
         }
 
         $builder = new McqValidationBuilder($question->get_multiple_responses(), $valueIdentifierMap);
-        list($responseDeclaration, $responseProcessing) = $builder->buildValidation($interactionIdentifier, $question->get_validation());
+        if (isset($feedbackOptions) && !empty($feedbackOptions)) {
+            list($responseDeclaration, $responseProcessing) = $builder->buildValidation($interactionIdentifier, $question->get_validation(), $feedbackOptions);
```
Comment (ttton) [inline src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php]: This feels a bit like unnecessary duplication - if we're always going to call the same method and the only difference is what is passed to `$feedbackOptions`, could you just always initialize `$feedbackOptions` to `[]` and remove the `else` branch below?

Looks like you already initialize it at the top of this `if`: (https://github.com/Learnosity/learnosity-qti/pull/34/files#diff-b83c6121e7dd5ee24648d67bf70bcf88R39)[https://github.com/Learnosity/learnosity-qti/pull/34/files#diff-b83c6121e7dd5ee24648d67bf70bcf88R39]. I think all you need to do is move it out of the `if`.

File: src/Processors/QtiV2/Out/ResponseProcessingBuilders/QtiResponseProcessingBuilder.php [hunk 17]
```diff
@@ -0,0 +1,53 @@
+<?php
+
+namespace LearnosityQti\Processors\QtiV2\Out\ResponseProcessingBuilders;
```
Comment (ttton) [inline src/Processors/QtiV2/Out/ResponseProcessingBuilders/QtiResponseProcessingBuilder.php]: I think it might be better to rename the namespace suffix to just `ResponseProcessing`. That makes it more extensible and adaptable later.

File: src/Processors/QtiV2/Out/AssessmentItemBuilder.php [hunk 18]
```diff
@@ -98,26 +115,32 @@ private function map(Question $question)
...
-        $interactionIdentifier = Format::isIdentifier($questionReference, false) ? $questionReference : strtoupper($type)  . '_' . StringUtil::generateRandomString(12);
-        if ($interactionIdentifier !== $questionReference) {
-            LogService::log(
-                "The question `reference` ($questionReference) is not a valid identifier. " .
-                "Replaced it with randomly generated `$interactionIdentifier` and stored the original `reference` as `label` attribute"
-            );
-        }
+        $interactionIdentifier = Format::isIdentifier($questionReference, false) ? $questionReference : strtoupper($type) . '_' . StringUtil::generateRandomString(12);
+
+        $interactionIdentifier = 'RESPONSE';
```
Comment (ttton) [inline src/Processors/QtiV2/Out/AssessmentItemBuilder.php]: Just noticed that this is rewriting the interaction identifier that is assigned immediately before it - and I imagine we don't want to have a non-unique identifier. Does this line just need to be removed?

File: src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php [hunk 19]
```diff
@@ -9,15 +9,21 @@
 use LearnosityQti\Entities\QuestionTypes\mcq_options_item;
 use LearnosityQti\Services\LogService;
 use LearnosityQti\Utils\QtiMarshallerUtil;
+use qtism\common\utils\Version;
```
[the remark this answers is not in the document]
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php]: We used it earlier but then we found some other work around and we had to remove that code in which it was used. 

Removed now

File: src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php [hunk 20]
```diff
@@ -26,11 +32,30 @@ public function convert(BaseQuestionType $questionType, $interactionIdentifier,
         // Build <choiceInteraction>
         $valueIdentifierMap = [];
         $simpleChoiceCollection = new SimpleChoiceCollection();
+        $inlineCollection = new InlineCollection();
+        $metadata = $question->get_metadata();
+
+        if (isset($metadata) && !empty($metadata->get_distractor_rationale_response_level())) {
+            $feedbackOptions = [];
+            foreach ($metadata->get_distractor_rationale_response_level() as $feed):
```
[the remark this answers is not in the document]
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php]: Done , We will keep this in mind in further PR.

File: src/Converter.php (hunk 12 above)
Comment (pooja31chaudhary) [reply src/Converter.php]: Yes , changed now.

File: src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php (hunk 13 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php]: I have removed all unused use statements.

File: src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php (hunk 14 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php]: Done

File: src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php (hunk 16 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php]: Done

File: src/Processors/QtiV2/Out/ResponseProcessingBuilders/QtiResponseProcessingBuilder.php (hunk 17 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/ResponseProcessingBuilders/QtiResponseProcessingBuilder.php]: Done

File: src/Processors/QtiV2/Out/AssessmentItemBuilder.php (hunk 18 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/AssessmentItemBuilder.php]: Actually earlier the converted qti.xml files were having more then one interaction in case of multiple questions in single item but now we have changed the flow and we are creating a separate xml file for each question in item json. So there would be only one interaction in a qti file and it is not an issue to have this as an interaction identifier.

File: src/Services/ConvertToQtiService.php (hunk 15 above)
Comment (pooja31chaudhary) [reply src/Services/ConvertToQtiService.php]: Done

File: src/Processors/QtiV2/Out/AssessmentItemBuilder.php (hunk 18 above)
Comment (ttton) [reply src/Processors/QtiV2/Out/AssessmentItemBuilder.php]: Ok, so my understanding of what you're saying here is that this line intentionally overrides the original behavior, which used the question reference to build the identifier.

If that's the case, can we remove the old logic? Having both here is confusing and redundant. Also update (or remove) any inline comments or docblock comments that describe the old behavior.

File: src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php [hunk 21]
```diff
@@ -26,13 +30,32 @@ public function convert(BaseQuestionType $questionType, $interactionIdentifier,
...
+        $metadata = $question->get_metadata();
+
+        $feedbackOptions = [];
+        if (isset($metadata) && !empty($metadata->get_distractor_rationale_response_level())) {
+            foreach ($metadata->get_distractor_rationale_response_level() as $feed) {
+                $feedbackOptions[] = $feed;
+            }
+        }
+
+        $i = 0;
```
Comment (ttton) [inline src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php]: Why not just use `$index` from inside the loop? Is `$index` a string key or an ordinal value? (if it's a string key then fair enough, but then `$index` should probably be called `$key` instead.)

Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php]: Done , sorry dint check $index . Now I have removed $i and using $index instead.

File: src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php [hunk 22]
```diff
@@ -26,13 +30,32 @@ public function convert(BaseQuestionType $questionType, $interactionIdentifier,
...
+
         foreach ($question->get_options() as $index => $option) {
+
             /** @var mcq_options_item $option */
             $choiceContent = new FlowStaticCollection();
             foreach (QtiMarshallerUtil::unmarshallElement($option->get_label()) as $component) {
+
                 $choiceContent->attach($component);
+                // attach feedbackInline to simpleChoice
+                if (isset($feedbackOptions) && $feedbackOptions[$index] !== '' && $component instanceof TextRun) {
```
[the remark this answers is not in the document]
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/QuestionTypes/McqMapper.php]: Done

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php [hunk 23]
```diff
@@ -12,7 +13,7 @@ abstract class AbstractQuestionValidationBuilder
 
     abstract protected function buildResponseDeclaration($responseIdentifier, $validation);
 
-    public function buildValidation($responseIdentifier, $validation, $isCaseSensitive = true)
+    public function buildValidation($responseIdentifier, $validation, $distractorRationaleResponseLevel = array(), $isCaseSensitive = true)
```
Comment (ttton) [inline src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: Sorry I didn't pick this up sooner - you've added a new optional parameter `$distractorRationaleResponseLevel` before the existing parameter `$isCaseSensitive`, which breaks the order of arguments. This will break calling classes that already pass `$isCaseSensitive`, unless we refactor the calls to match. See the `shorttext` question type here: https://github.com/Learnosity/learnosity-qti/blob/develop/src/Processors/QtiV2/Out/QuestionTypes/ShorttextMapper.php#L42

Be careful of making breaking changes like this - you'll either have to find and refactor all method calls, or add the new parameters at the end.

Comment (rakeshqdsa) [reply src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: Done
